src/context/online.rs

```diff
@@ -140,15 +165,73 @@ impl AudioContext {
     /// the time at which the first sample in the buffer is actually processed
     /// by the audio output device.
     #[must_use]
+    #[allow(clippy::missing_panics_doc)]
     pub fn output_latency(&self) -> f64 {
-        self.backend.output_latency()
+        self.backend.lock().unwrap().output_latency()
     }
 
     /// Identifier or the information of the current audio output device.
     ///
     /// The initial value is `None`, which means the default audio output device.
-    pub fn sink_id(&self) -> Option<&str> {
-        self.backend.sink_id()
+    #[allow(clippy::missing_panics_doc)]
+    pub fn sink_id(&self) -> Option<String> {
+        self.backend.lock().unwrap().sink_id().map(str::to_string)
+    }
+
+    /// Update the current audio output device.
+    ///
+    /// This function operates synchronously and might block the current thread. An async version
+    /// is currently not implemented.
+    #[allow(clippy::needless_collect, clippy::missing_panics_doc)]
+    pub fn set_sink_id_sync(&self, sink_id: String) -> Result<(), Box<dyn Error>> {
+        if self.sink_id().as_deref() == Some(&sink_id) {
+            return Ok(()); // sink is already active
+        }
+
+        if !crate::enumerate_devices()
+            .into_iter()
+            .any(|d| d.device_id() == sink_id)
+        {
+            Err("NotFoundError: invalid sinkId")?;
+        }
+
+        let mut backend_guard = self.backend.lock().unwrap();
+
+        // acquire exclusive lock on ctrl msg sender
+        let ctrl_msg_send = self.base.lock_control_msg_sender();
+
+        // flush out the ctrl msg receiver, cache
+        let pending_msgs: Vec<_> = self.render_thread_init.ctrl_msg_recv.try_iter().collect();
+
+        // acquire the audio graph from the current render thread, shutting it down
+        let (graph_send, graph_recv) = crossbeam_channel::bounded(1);
+        let message = ControlMessage::Shutdown { sender: graph_send };
+        ctrl_msg_send.send(message).unwrap();
+        let graph = graph_recv.recv().unwrap();
+
+        // hotswap the backend
+        let options = AudioContextOptions {
+            sample_rate: Some(self.sample_rate()),
+            latency_hint: AudioContextLatencyCategory::default(), // todo reuse existing setting
+            sink_id: Some(sink_id),
+        };
+        *backend_guard = io::build_output(options, self.render_thread_init.clone());
+
+        // send the audio graph to the new render thread
+        let message = ControlMessage::Startup { graph };
+        ctrl_msg_send.send(message).unwrap();
+
+        self.base().set_state(AudioContextState::Running);
```

Comment:
if the context was suspended maybe we want to keep it suspended?

Actually, I didn't read the thing carefully, but the spec seems to say:

>If wasRunning is true:
>
> Set the [[[rendering thread state]]](https://webaudio.github.io/web-audio-api/#dom-baseaudiocontext-rendering-thread-state-slot) on the [AudioContext](https://webaudio.github.io/web-audio-api/#AudioContext) to "suspended".

https://webaudio.github.io/web-audio-api/#ref-for-dom-audiocontext-setsinkid%E2%91%A0

But as we already differ on the initial state, I don't really know... 
Just keeping the state as it was could be a good tradeoff?

Reply:
Yes, very good point. I will update the PR